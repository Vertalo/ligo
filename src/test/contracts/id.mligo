type id = string

type id_details = {
  owner: address;
  controller: address;
  profile: bytes;
}

type storage = (id, id_details) big_map * int

(* TODO: Figure out how to keep this price reasonable even as cost of tz fluctuates *)
let price = 10tz

(** Preliminary thoughts on ids:

I very much like the simplicity of http://gurno.com/adam/mne/.
5 three letter words means you have a 15 character identity, not actually more
annoying than an IP address and a lot more memorable than the raw digits. This
can be stored as a single integer which is then translated into the corresponding
series of 5 words.

I in general like the idea of having a 'skip' mechanism, but it does need to cost
something so people don't eat up the address space. 256 ^ 5 means you have a lot
of address space, but if people troll by skipping a lot that could be eaten up.
Should probably do some napkin calculations for how expensive skipping needs to
be to deter people from doing it just to chew up address space.
*)  

let buy (parameter: (bytes * address option) * storage) =
  assert (amount = price) ;
  let profile: bytes = parameter.0.0 in
  let controller: address =
    match parameter.0.1 with
    | Some addr -> addr
    | None -> sender
  in
  let new_id: id = parameter.1.1 + 1 in
  let untaken: bool =
    match Big_map.find_opt new_name parameter.1.0 with
    | Some name -> failwith("The name " ^ new_id ^ "has already been registered.") 
    | None -> true
  in
  let new_id_details: id_details = {
    owner = sender ;
    controller = controller ;
    profile = profile ;
  }
  in
  let updated_identities: (id, id_details) big_map =
    Big_map.update new_id new_id_details parameter.1.0
  in
  in ([]: instruction, (updated_identities, new_id))

let update_owner (parameter: (id * address) * storage) =
  let (id: id), (new_owner: address) = paramater.0 in
  let identities = parameter.1.0 in
  let current_id_details = Bip_map.find_opt id identities in
  let is_allowed: bool =
    if sender = current_id_details.owner
    then true
    else failwith "You are not the owner of the ID " ^ id
  in
  let updated_id_details = {
    owner = new_owner;
    controller = current_id_details.controller;
    profile = current_id_details.profile;
  }
  in
  let updated_identities = Big_map.update id updated_id_details identities in
  ([]: instruction, updated_identities)

let update_details (parameter: (id * bytes * address option) * storage) =
  let current_id_details = Big_map.find_opt parameter.0.0 parameter.1.0 in
  let is_allowed: bool =
    if
      match current_id_details with
      | Some id_details -> (sender = id_details.controller) || (sender = id_details.owner)
      | None -> failwith ("No such ID " + parameter.0.0)
    then true
    else failwith ("You are not the owner or controller of the ID " ^ parameter.0.0)
  in
  let identities: (id, id_details) big_map = parameter.1.0 in
  let (id: id), (profile: bytes), (controller: address) = parameter.0 in
  let owner: address = current_id_details.owner in
  let controller: address =
    match parameter.0.2 with
    | None -> (* Default *) current_id_details.controller
    | Some new_controller -> new_controller
  in
  let updated_id_details = {
    owner = owner;
    controller = controller;
    profile = profile;
  }
  let updated_identities = Big_map.update id updated_id_details identities in
  ([]: instruction, updated_identities)

(* Let someone skip the next identity so nobody has to take one that's undesirable *)
let skip (p: unit) = ()

let whois_id (parameter: id * storage) =
  let query: id = parameter.0 in
  let identities: (id, id_details) big_map = parameter.1.0 in
  let result: id_details =
    match Big_map.find_opt query identities with
    | Some id_details -> id_details
    | None -> failwith "This ID doesn't exist in the system."
  in () (* Not entirely sure how to return the result here *)
