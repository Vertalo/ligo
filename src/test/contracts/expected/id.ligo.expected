type id is int

type id_details is
  record [
    owner : address;
    controller : address;
    profile : bytes
  ]

type buy is
  record [
    profile : bytes;
    initial_controller : option (address)
  ]

type update_owner is record [id : id; new_owner : address]

type update_details is
  record [
    id : id;
    new_profile : option (bytes);
    new_controller : option (address)
  ]

type action is
    Buy of buy | Update_owner of update_owner
  | Update_details of update_details | Skip of unit

type storage is
  record [
    identities : big_map (id, id_details);
    next_id : int;
    name_price : tez;
    skip_price : tez
  ]

function buy
  (const parameter : buy;
   const storage : storage) : list (operation) * storage is
block {
  if amount = storage.name_price
  then skip
  else failwith ("Incorrect amount paid.");
  const profile : bytes = parameter.profile;
  const initial_controller : option (address)
  = parameter.initial_controller;
  var identities : big_map (id, id_details)
  := storage.identities;
  const new_id : int = storage.next_id;
  const controller : address
  = case initial_controller of [
      Some (addr) -> addr
    | None -> sender
    ];
  const new_id_details : id_details
  = record [
      owner = sender;
      controller = controller;
      profile = profile
    ];
  identities [new_id] := new_id_details
} with
    ((nil : list (operation)),
     storage with
       record [identities = identities; next_id = new_id + 1
       ])

function update_owner
  (const parameter : update_owner;
   const storage : storage) : list (operation) * storage is
block {
  if (amount =/= 0mutez)
  then
    block {
      failwith ("Updating owner doesn't cost anything.")
    }
  else skip;
  const id : int = parameter.id;
  const new_owner : address = parameter.new_owner;
  var identities : big_map (id, id_details)
  := storage.identities;
  const id_details : id_details
  = case identities [id] of [
      Some (id_details) -> id_details
    | None ->
        (failwith ("This ID does not exist.") : id_details)
    ];
  if sender = id_details.owner
  then skip
  else failwith ("You are not the owner of this ID.");
  id_details.owner := new_owner;
  identities [id] := id_details
} with
    ((nil : list (operation)),
     storage with
       record [identities = identities])

function update_details
  (const parameter : update_details;
   const storage : storage) : list (operation) * storage is
block {
  if (amount =/= 0mutez)
  then failwith ("Updating details doesn't cost anything.")
  else skip;
  const id : int = parameter.id;
  const new_profile : option (bytes) = parameter.new_profile;
  const new_controller : option (address)
  = parameter.new_controller;
  const identities : big_map (id, id_details)
  = storage.identities;
  const id_details : id_details
  = case identities [id] of [
      Some (id_details) -> id_details
    | None ->
        (failwith ("This ID does not exist.") : id_details)
    ];
  if (sender = id_details.controller)
     or (sender = id_details.owner)
  then skip
  else
    failwith
      ("You are not the owner or controller of this ID.");
  const owner : address = id_details.owner;
  const profile : bytes
  = case new_profile of [
      None -> id_details.profile
    | Some (new_profile) -> new_profile
    ];
  const controller : address
  = case new_controller of [
      None -> id_details.controller
    | Some (new_controller) -> new_controller
    ];
  id_details.owner := owner;
  id_details.controller := controller;
  id_details.profile := profile;
  identities [id] := id_details
} with
    ((nil : list (operation)),
     storage with
       record [identities = identities])

function skip_ (const p : unit; const storage : storage)
  : list (operation) * storage is
block {
  if amount = storage.skip_price
  then skip
  else failwith ("Incorrect amount paid.")
} with
    ((nil : list (operation)),
     storage with
       record [next_id = storage.next_id + 1])

function main
  (const action : action;
   const storage : storage) : list (operation) * storage is
  case action of [
    Buy (b) -> buy (b, storage)
  | Update_owner (uo) -> update_owner (uo, storage)
  | Update_details (ud) -> update_details (ud, storage)
  | Skip (s) -> skip_ (unit, storage)
  ]
